Branch xingyu dg contribution (#194)
AY2021S1-CS2103T-T12-1/tp: docs/DeveloperGuide.md

This section describes some noteworthy details on how certain features are implemented.

### Adding Items and Delivery
OneShelf is capable of adding items and deliveries.
Adding Items and Delivery both are done similarly which will be illustrated with an AddItemActivityDiagram below.

![AddItemActivityDiagram](images/AddItemActivityDiagram.png)

> We've repurposed the `Add` command from AB3 to add on to existing item's quantity instead of throwing an error.

Reviewer: ```suggestion
Apart from adding a new Item, should there be an existing item in the Inventory Book, `Add` will be able to increase the quantity of that particular Item.
```

In this case 2 items are considered the same if they have the same `Name` and `Supplier`.
The other variables such as:

a) `Metric` `MaxQuantity`

> - Are not allowed to be defined if there exist the same item inside InventoryBook.

Reviewer: ```suggestion
- Are not allowed to be defined if there exists the same item inside InventoryBook.
```

b) `Tags`

> - Will be combined together if there exist the same item inside InventoryBook.

Reviewer: ```suggestion
- Will be combined together if there exists the same item inside InventoryBook.
```

Author: Thanks! Sorry for the oversight

> Note: No 2 deliveries are considered the same. Reason being the same person can make multiple delivery orders.

Reviewer: No 2 can sound like number 2. And I don't really understand what is the message you're trying to bring across, perhaps a paraphrasing is needed?

Author: Paraphrased!

This section describes some noteworthy details on how certain features are implemented.

### Adding Items and Delivery
OneShelf is capable of adding items and deliveries.
Adding Items and Delivery both are done similarly which will be illustrated with an AddItemActivityDiagram below.

![AddItemActivityDiagram](images/AddItemActivityDiagram.png)

Apart from adding a new Item, should there be an existing item in the Inventory Book, `Add` will be able to increase the quantity of that particular Item.
In this case 2 items are considered the same if they have the same `Name` and `Supplier`.
The other variables such as:

a) `Metric` `MaxQuantity`
- Are not allowed to be defined if there exists the same item inside InventoryBook.

b) `Tags`
- Will be combined together if there exist the same item inside InventoryBook.

Note: Deliveries are all considered unique. Reason being the same person can make multiple delivery orders.

> For example, `John` with the address `Choa Chu Kang Block 259` is able to make multiple orders before his previous deliveries are fulfilled.

Reviewer: U can use the information source for this kind of note to make it consistent

Author: Added! Thanks!

With `addToHistory(String command)`, `previousCommand()`, `nextCommand()` and `currentCommand()` implemented, a simple `setOnKeyPressed` under `CommandBox` class which checks
for user's input of arrow up (which calls previousCommand()) and arrow down (which calls nextCommand()) would suffice for GUI implementation.

Below is the sequence diagram when user pressing the arrow up button with `CommandBox` selected on GUI.

![CommandHistoryTraversalSequenceDiagram](images/CommandHistoryTraversalSequenceDiagram.png)

> The previous command by the user will be reflected in the GUI.

Reviewer: Would you want to explain or provide a brief explanation of the diagram?

Author: Added!

1. _{ more test cases …​ }_

### Adding an item

> 1. Adding an item

Reviewer: I think it'll be even better if you split up into 2 sections in this adding part: Adding to an existing item, adding to a non-existing item (instead of explaining them in the expected part)

Author: Done! That's a good idea thank you!

1. _{ more test cases …​ }_

if there is no existing same item. If there is an existing same item, 

### Adding an item

1. Adding an item
    1. Test Case: `add-i n/Chicken q/123 s/NTUC`
       Expected: Item with `Name` of Chicken, `Quantity` of 123 and `Supplier` of NTUC added 
       
    1. Test Case: `add-i n/Chicken q/123 s/giant max/500 metric/kg`

>        Expected: Item with `Name` of Chicken, `Quantity` of 123, `Supplier` of NTUC, `MaxQuantity` of 500 and `Metric` of kg added if there is no existing same item. 

Reviewer: ```suggestion
       Expected: Item with `Name` of Chicken, `Quantity` of 123, `Supplier` of NTUC, `MaxQuantity` of 500 and `Metric` of kg added when there is no existing same item. 
```

Author: Remedied! Good catch!
